packages/components/src/public/files/TemplateDownloadButton.tsx

```diff
@@ -1,155 +1,184 @@
-    }, [customTemplates, isGridRenderer]);
+        return hasTemplates_;
+    }, [api.query, container, moduleContext, schemaName, queryName]);
+
+    const downloadCustomTemplate = useCallback(
+        (customTemplateUrl: string) => {
+            return async () => {
+                if (downloading) return;
+                setDownloading(true);
+                await downloadAttachment(customTemplateUrl);
+                setDownloading(false);
+            };
+        },
+        [downloading]
+    );
+
+    const downloadDefaultTemplate = useCallback(async () => {
+        if (downloading) return;
+        setDownloading(true);
+        if (onDownloadDefault) {
+            await onDownloadDefault();
+        } else {
+            await downloadAttachment(defaultTemplateUrl);
+        }
+        setDownloading(false);
+    }, [defaultTemplateUrl, downloading, onDownloadDefault]);
 
     const fetchTemplates = useCallback(async () => {
-        if (customTemplates || !isGridRenderer) return;
-        const hasCustomTemplates = await loadTemplates();
+        if (isDownloadingOrLoading || !schemaQuery) return;
+        let hasCustomTemplates: boolean;
+
+        if (isLoaded) {
+            hasCustomTemplates = hasTemplates;
+        } else {
+            hasCustomTemplates = await loadCustomTemplates();
+        }
+
+        // There are no custom templates so download the default template
         if (!hasCustomTemplates) {
-            if (onDownloadDefault) onDownloadDefault();
-            else downloadAttachment(defaultTemplateUrl, true);
+            downloadDefaultTemplate();
         }
-    }, [isGridRenderer, onDownloadDefault, customTemplates, loadTemplates, defaultTemplateUrl]);
+    }, [downloadDefaultTemplate, hasTemplates, isDownloadingOrLoading, isLoaded, loadCustomTemplates, schemaQuery]);
+
+    const buttonClassName = classNames('template-download-button', {
+        'button-small-padding': isGridRenderer,
+
+        // marker classes to assist automated testing
+        'has-templates': hasTemplates,
+        'is-loaded': isLoaded,
+    });
 
     const dropdownTitle = useMemo(() => {
+        const iconClassName = classNames('fa', {
+            'fa-download': !isDownloadingOrLoading,
+            'fa-spinner': isDownloadingOrLoading,
+            'fa-pulse': isDownloadingOrLoading,
+        });
+
         return (
-            <span title="Download Template">
-                <span className="fa fa-download" /> {text}
+            <span title={TITLE}>
+                <span className={iconClassName} /> {text}
             </span>
         );
-    }, [text]);
+    }, [isDownloadingOrLoading, text]);
 
-    if (!onDownloadDefault && !defaultTemplateUrl?.length && !showDropdown) return null;
+    return (
+        <DropdownButton
+            bsStyle="info"
+            buttonClassName={buttonClassName}
+            buttonTitle={TITLE}
+            className={className}
+            noCaret
```

Comment:
Is it intentional that this is no always noCaret?

Reply:
Yes, having the button jump around due to the presence of menu didn't seem like a good user experience and the menu appears if needed. The other choice is to always show a caret and then don't have a menu appear sometimes.

Reply:
That might be the case of the data types listing grid. But for other places, like listing page or import page, the menu/button can be rendered correctly on loading. It seems useful to at least show dropdown in those places.